ngraph/test/backend/unary_test.hpp

```diff
@@ -0,0 +1,211 @@
+    // threshold), corresponding to two different comparing methods implemented in TestCase
+    // classes. Following class acts as an encapsulation on these two options without introducing
+    // too many overloadings on the function requiring such argument.
+    struct ArgTolerance
+    {
+        const int tolerance_bits;
+        const float tolerance_fp;
+        bool using_bits;
+        ArgTolerance()
+            : tolerance_bits(-1)
+            , tolerance_fp(0)
+            , using_bits(true)
+        {
+        }
+        ArgTolerance(const int tolerance_bits)
+            : tolerance_bits(tolerance_bits)
+            , tolerance_fp(0)
+            , using_bits(true)
+        {
+        }
+        ArgTolerance(const float tolerance_fp)
+            : tolerance_bits(-1)
+            , tolerance_fp(tolerance_fp)
+            , using_bits(false)
+        {
+        }
+        ArgTolerance(const double tolerance_fp)
+            : tolerance_bits(-1)
+            , tolerance_fp(tolerance_fp)
+            , using_bits(false)
+        {
+        }
+    };
+
+    // ArgShape is used implicitly by caller of test_unary() in following ways:
+    //
+    //  {PartialShape{2,Dimension::dynamic(),3}, Shape{2, 4, 3}}
+    //        A full specified form with both partial shape for ngraph function
+    //        generation and a static shape for final input tensor.
+    //
+    //  Shape{2, 4, 3}
+    //        A pure static shape for both ngraph function generation and input tensor
+    //
+    //  {}
+    //        A unspecified shape, test_unary() will derive the actual shape from the size
+    //        of the input vector as 1D shape.
+    struct ArgShape
+    {
+        PartialShape dynamic_shape;
+        Shape static_shape;
+        bool initialzed;
+        ArgShape(const PartialShape& dynamic_shape, const Shape& static_shape)
+            : dynamic_shape(dynamic_shape)
+            , static_shape(static_shape)
+            , initialzed(true)
+        {
+        }
+        ArgShape(const Shape& static_shape)
+            : dynamic_shape(static_shape)
+            , static_shape(static_shape)
+            , initialzed(true)
+        {
+        }
+        ArgShape()
+            : dynamic_shape{}
+            , static_shape{}
+            , initialzed(false)
+        {
+        }
+    };
+
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(std::shared_ptr<Function> f,
+                    const std::vector<value_type>& test_input,
+                    const std::vector<value_type>& test_expected,
+                    ArgShape ashape = {},
+                    ArgTolerance tol = {})
+    {
+        if (!ashape.initialzed)
+        {
+            ashape = ArgShape(Shape({test_input.size()}));
+        }
+        // the size specified in static_shape may be bigger or smaller than the
```

Comment:
What's the use case here? In all tests we always had a very simple approach - whether you're passing input or expected output values, they have to match the input or output shape. The sampling and modulo operator for indexing looks a little hacky. What's the purpose?

Reply:
Suppose we want to test on a 4D input of shape{2,3,4,5} requiring 120 const data for inputs and expected results, which would be too much to write them down in source code, especially when the point of the test is to check if the Op is working well on 4D shapes rather than on 120 different specific values, in that case we can just pass much fewer number of inputs (<<120) and test_unary() can help us to expand them to required shape.

Reply:
In this case I would suggest feeding random numbers and creating a function that produces correct reference values based on the same random input data. TestCase also allows reading values in from a file. If I understand it correctly, when the shape is {2,3,4,5} and you only provide a few numbers(less than 120) they get copied multiple times into the final vector containing input or expected output. If this is the behavior, please adjust the comment to reflect that.

Reply:
OK